### PR TITLE
feat(lpa-questionnaire-web-app): add validation for development plan

### PR DESCRIPTION
**Pull request by swdpcomputing**
## Ticket Number
AS-1302

## Description of change
add validation for development plan

## Checklist

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `master` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

File: packages/lpa-questionnaire-web-app/src/validators/development-plan.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+const rules = () => {
+  const developmentPlanRef = 'has-plan-submitted';
+  const developmentPlanValues = ['yes', 'no'];
+  const developmentPlanTextRef = 'plan-changes-text';
+
+  return [
+    body(developmentPlanRef)
+      .notEmpty()
+      .withMessage(
+        'Has any Development Plan Document or Neighbourhood Plan relevant to this appeal been submitted for examination (or in the case of a Neighbourhood Plan, been examined and is awaiting a referendum)?'
```
Comment (davehlewis) [inline packages/lpa-questionnaire-web-app/src/validators/development-plan.js:12]: This is not the correct error. Should be 'Select yes if there is a relevant Development Plan or Neighbourhood Plan'

File: packages/lpa-questionnaire-web-app/src/validators/development-plan.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      )
+      .bail()
+      .isIn(developmentPlanValues),
+    body(developmentPlanTextRef)
+      .if(body(developmentPlanRef).equals('yes'))
+      .notEmpty()
+      .withMessage(
+        'Tell us about any substantive changes in the progress of the emerging plan, and their relevance to this appeal (if the plan may not be adopted before the Inspector issues their decision).'
```
Comment (davehlewis) [inline packages/lpa-questionnaire-web-app/src/validators/development-plan.js:20]: This is not the correct error. Should be 'Enter the relevant information about the plans and this appeal'. Need to check these from the ticket.

File: packages/lpa-questionnaire-web-app/tests/unit/validators/development-plan.test.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+      expect(rule.fields).toEqual(['has-plan-submitted']);
+      expect(rule.locations).toEqual(['body']);
+      expect(rule.optional).toBeFalsy();
+      expect(rule.stack).toHaveLength(3);
+
+      expect(rule.stack[0].validator.name).toEqual('isEmpty');
+      expect(rule.stack[0].negated).toBeTruthy();
+      expect(rule.stack[0].message).toEqual(
+        'Has any Development Plan Document or Neighbourhood Plan relevant to this appeal been submitted for examination (or in the case of a Neighbourhood Plan, been examined and is awaiting a referendum)?'
```
Comment (davehlewis) [inline packages/lpa-questionnaire-web-app/tests/unit/validators/development-plan.test.js:18]: Needs fixing as above

File: packages/lpa-questionnaire-web-app/tests/unit/validators/development-plan.test.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      expect(rule.stack).toHaveLength(2);
+
+      expect(rule.stack[0].chain).toHaveProperty('if');
+      expect(rule.stack[0].chain.builder.fields).toEqual(['has-plan-submitted']);
+      expect(rule.stack[0].chain.builder.locations).toEqual(['body']);
+
+      expect(rule.stack[1].validator.name).toEqual('isEmpty');
+      expect(rule.stack[1].negated).toBeTruthy();
+      expect(rule.stack[1].message).toEqual(
+        'Tell us about any substantive changes in the progress of the emerging plan, and their relevance to this appeal (if the plan may not be adopted before the Inspector issues their decision).'
```
Comment (davehlewis) [inline packages/lpa-questionnaire-web-app/tests/unit/validators/development-plan.test.js:40]: As above

**Review by davehlewis — CHANGES_REQUESTED** [review]
Looks like there are some missing references, as well as steps from e2e testing that need fixing.

Also validator is not attached to component at all. Shouldn't that be part of this?

File: packages/lpa-questionnaire-web-app/src/validators/development-plan.js (hunk 1 above)
Comment (swdpcomputing) [reply packages/lpa-questionnaire-web-app/src/validators/development-plan.js:12]: Corrected

File: packages/lpa-questionnaire-web-app/src/validators/development-plan.js (hunk 2 above)
Comment (swdpcomputing) [reply packages/lpa-questionnaire-web-app/src/validators/development-plan.js:20]: Corrected

File: packages/lpa-questionnaire-web-app/tests/unit/validators/development-plan.test.js (hunk 3 above)
Comment (swdpcomputing) [reply packages/lpa-questionnaire-web-app/tests/unit/validators/development-plan.test.js:18]: Corrected

File: packages/lpa-questionnaire-web-app/tests/unit/validators/development-plan.test.js (hunk 4 above)
Comment (swdpcomputing) [reply packages/lpa-questionnaire-web-app/tests/unit/validators/development-plan.test.js:40]: Corrected

**Comment by swdpcomputing** [comment]
> Looks like there are some missing references, as well as steps from e2e testing that need fixing.
> 
> Also validator is not attached to component at all. Shouldn't that be part of this?

Fixed e2e. Marc and I found that doing the POST request as part of this ticket just to get the validator in was not the best way. The validator will be used in the next ticket.

**Review by davehlewis — APPROVED** [review]

**Pull request merged**